File: backend/app/schemas/maintenance.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field

from app.schemas.common import EntityBase, SoftDeleteFields


# ---------------------------------------------------------------------------
# Dictionary schemas (shared pattern)
# ---------------------------------------------------------------------------

class MntDictBase(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    code: str | None = Field(default=None, max_length=50)
    description: str | None = None


class MntDictOut(EntityBase, SoftDeleteFields):
    name: str
    code: str | None = None
    description: str | None = None


# --- Failure Modes ---

class FailureModeCreate(MntDictBase):
    equipment_category_id: int | None = None
    parent_id: int | None = None


class FailureModeUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    code: str | None = Field(default=None, max_length=50)
    description: str | None = None
    equipment_category_id: int | None = None
    parent_id: int | None = None
    is_deleted: bool | None = None


class FailureModeOut(MntDictOut):
    equipment_category_id: int | None = None
    parent_id: int | None = None


# --- Failure Mechanisms ---

class FailureMechanismCreate(MntDictBase):
    pass


class FailureMechanismUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    code: str | None = Field(default=None, max_length=50)
    description: str | None = None
    is_deleted: bool | None = None


class FailureMechanismOut(MntDictOut):
    pass


# --- Failure Causes ---

class FailureCauseCreate(MntDictBase):
    category: str | None = Field(default=None, max_length=50)


class FailureCauseUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    code: str | None = Field(default=None, max_length=50)
    category: str | None = Field(default=None, max_length=50)
    description: str | None = None
    is_deleted: bool | None = None


class FailureCauseOut(MntDictOut):
    category: str | None = None


# --- Detection Methods ---

class DetectionMethodCreate(MntDictBase):
    pass


class DetectionMethodUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    code: str | None = Field(default=None, max_length=50)
    description: str | None = None
    is_deleted: bool | None = None


class DetectionMethodOut(MntDictOut):
    pass


# --- Activity Types ---

class ActivityTypeCreate(MntDictBase):
    category: str | None = Field(default=None, max_length=50)


class ActivityTypeUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    code: str | None = Field(default=None, max_length=50)
    category: str | None = Field(default=None, max_length=50)
    description: str | None = None
    is_deleted: bool | None = None


class ActivityTypeOut(MntDictOut):
    category: str | None = None


# ---------------------------------------------------------------------------
# Incident Component
# ---------------------------------------------------------------------------

class IncidentComponentCreate(BaseModel):
    cabinet_item_id: int
    equipment_type_id: int | None = None
    failure_mode_id: int | None = None
    damage_description: str | None = None
    action_taken: str | None = Field(default=None, max_length=50)


class IncidentComponentOut(BaseModel):
    id: int
    incident_id: int
    cabinet_item_id: int
    equipment_type_id: int | None = None
    failure_mode_id: int | None = None
    damage_description: str | None = None
    action_taken: str | None = None
    # joined names
    equipment_type_name: str | None = None
    failure_mode_name: str | None = None
    created_at: datetime


# ---------------------------------------------------------------------------
# Incidents
# ---------------------------------------------------------------------------

class IncidentCreate(BaseModel):
    cabinet_id: int
    location_id: int | None = None
    severity: str | None = Field(default=None, max_length=20)
    detection_method_id: int | None = None
    failure_mode_id: int | None = None
    failure_mechanism_id: int | None = None
    failure_cause_id: int | None = None
    occurred_at: datetime
    detected_at: datetime
    title: str = Field(min_length=1, max_length=200)
    description: str | None = None
    operational_impact: str | None = Field(default=None, max_length=32)
    assigned_to_id: int | None = None
    components: list[IncidentComponentCreate] | None = None


class IncidentUpdate(BaseModel):
    severity: str | None = Field(default=None, max_length=20)
    status: str | None = Field(default=None, max_length=20)
    detection_method_id: int | None = None
    failure_mode_id: int | None = None
    failure_mechanism_id: int | None = None
    failure_cause_id: int | None = None
    repair_started_at: datetime | None = None
    resolved_at: datetime | None = None
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    root_cause_analysis: str | None = None
    resolution_notes: str | None = None
    man_hours: float | None = None
    downtime_hours: float | None = None
    operational_impact: str | None = Field(default=None, max_length=32)
    assigned_to_id: int | None = None
    is_deleted: bool | None = None


class IncidentOut(EntityBase, SoftDeleteFields):
    incident_number: str | None = None
    cabinet_id: int
    location_id: int | None = None
    severity: str | None = None
    detection_method_id: int | None = None
    failure_mode_id: int | None = None
    failure_mechanism_id: int | None = None
    failure_cause_id: int | None = None
    status: str
    occurred_at: datetime
    detected_at: datetime
    repair_started_at: datetime | None = None
    resolved_at: datetime | None = None
    title: str
    description: str | None = None
    root_cause_analysis: str | None = None
    resolution_notes: str | None = None
    man_hours: float | None = None
    downtime_hours: float | None = None
    operational_impact: str | None = None
    reported_by_id: int
    assigned_to_id: int | None = None
    # joined
    cabinet_name: str | None = None
    reported_by_username: str | None = None
    failure_mode_name: str | None = None
    failure_mechanism_name: str | None = None
    failure_cause_name: str | None = None
    detection_method_name: str | None = None


# ---------------------------------------------------------------------------
# Work Order Items
# ---------------------------------------------------------------------------

class WorkOrderItemCreate(BaseModel):
    cabinet_item_id: int | None = None
    equipment_type_id: int | None = None
    action: str = Field(max_length=20)
    quantity: int = 1
    notes: str | None = None


class WorkOrderItemOut(BaseModel):
    id: int
    work_order_id: int
    cabinet_item_id: int | None = None
    equipment_type_id: int | None = None
    action: str
    quantity: int
    notes: str | None = None
    equipment_type_name: str | None = None
    created_at: datetime


# ---------------------------------------------------------------------------
# Work Orders
# ---------------------------------------------------------------------------

class WorkOrderCreate(BaseModel):
    order_type: str = Field(max_length=32)
    activity_type_id: int | None = None
    priority: str = Field(default="normal", max_length=20)
    cabinet_id: int | None = None
    incident_id: int | None = None
    plan_id: int | None = None
    planned_start_date: date | None = None
    planned_end_date: date | None = None
    estimated_man_hours: float | None = None
    assigned_to_id: int | None = None
    title: str = Field(min_length=1, max_length=200)
    description: str | None = None
    items: list[WorkOrderItemCreate] | None = None


class WorkOrderUpdate(BaseModel):
    order_type: str | None = Field(default=None, max_length=32)
    activity_type_id: int | None = None
    priority: str | None = Field(default=None, max_length=20)
    status: str | None = Field(default=None, max_length=20)
    cabinet_id: int | None = None
    planned_start_date: date | None = None
    planned_end_date: date | None = None
    actual_start_at: datetime | None = None
    actual_end_at: datetime | None = None
    estimated_man_hours: float | None = None
    actual_man_hours: float | None = None
    assigned_to_id: int | None = None
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    completion_notes: str | None = None
    is_deleted: bool | None = None


class WorkOrderOut(EntityBase, SoftDeleteFields):
    order_number: str | None = None
    order_type: str
    activity_type_id: int | None = None
    priority: str
    status: str
    cabinet_id: int | None = None
    incident_id: int | None = None
    plan_id: int | None = None
    planned_start_date: date | None = None
    planned_end_date: date | None = None
    actual_start_at: datetime | None = None
    actual_end_at: datetime | None = None
    estimated_man_hours: float | None = None
    actual_man_hours: float | None = None
    assigned_to_id: int | None = None
    performed_by_id: int
    title: str
    description: str | None = None
    completion_notes: str | None = None
    # joined
    cabinet_name: str | None = None
    activity_type_name: str | None = None


# ---------------------------------------------------------------------------
# Maintenance Plans
# ---------------------------------------------------------------------------

class PlanCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    code: str | None = Field(default=None, max_length=50)
    equipment_category_id: int | None = None
    equipment_type_id: int | None = None
    cabinet_id: int | None = None
    interval_days: int = Field(gt=0)
    activity_type_id: int | None = None
    estimated_man_hours: float | None = None
    description: str | None = None
    next_due_date: date | None = None


class PlanUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    code: str | None = Field(default=None, max_length=50)
    equipment_category_id: int | None = None
    equipment_type_id: int | None = None
    cabinet_id: int | None = None
    interval_days: int | None = Field(default=None, gt=0)
    activity_type_id: int | None = None
    estimated_man_hours: float | None = None
    description: str | None = None
    next_due_date: date | None = None
    is_deleted: bool | None = None


class PlanOut(EntityBase, SoftDeleteFields):
    name: str
    code: str | None = None
    equipment_category_id: int | None = None
    equipment_type_id: int | None = None
    cabinet_id: int | None = None
    interval_days: int
    activity_type_id: int | None = None
    estimated_man_hours: float | None = None
    description: str | None = None
    last_generated_date: date | None = None
    next_due_date: date | None = None
    # joined
    cabinet_name: str | None = None
    activity_type_name: str | None = None


# ---------------------------------------------------------------------------
# Operating Time
# ---------------------------------------------------------------------------

class OperatingTimeCreate(BaseModel):
    cabinet_id: int
    recorded_date: date
    operating_hours: float = Field(ge=0, le=24)
    standby_hours: float = Field(default=0, ge=0, le=24)
    downtime_hours: float = Field(default=0, ge=0, le=24)
    notes: str | None = None


class OperatingTimeUpdate(BaseModel):
    operating_hours: float | None = Field(default=None, ge=0, le=24)
    standby_hours: float | None = Field(default=None, ge=0, le=24)
    downtime_hours: float | None = Field(default=None, ge=0, le=24)
    notes: str | None = None


class OperatingTimeOut(BaseModel):
    id: int
    cabinet_id: int
    recorded_date: date
    operating_hours: float
    standby_hours: float
    downtime_hours: float
    recorded_by_id: int
    notes: str | None = None
    created_at: datetime
    updated_at: datetime
    cabinet_name: str | None = None


# ---------------------------------------------------------------------------
# Reliability analytics
# ---------------------------------------------------------------------------

class ReliabilitySummary(BaseModel):
    cabinet_id: int | None = None
    cabinet_name: str | None = None
    total_incidents: int = 0
    total_operating_hours: float = 0
    total_downtime_hours: float = 0
    mtbf_hours: float | None = None
    mttr_hours: float | None = None
    availability_pct: float | None = None


class FailureTrendPoint(BaseModel):
    period: str
    incident_count: int


class TopFailure(BaseModel):
    equipment_type_id: int
    equipment_type_name: str | None = None
    incident_count: int

File: frontend/src/api/maintenance.ts
import { apiFetch } from "./client";
import { type Pagination, buildQuery } from "./entities";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type MntDict = {
  id: number;
  name: string;
  code: string | null;
  description: string | null;
  category?: string | null;
  equipment_category_id?: number | null;
  parent_id?: number | null;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
};

export type MntIncident = {
  id: number;
  incident_number: string | null;
  cabinet_id: number;
  location_id: number | null;
  severity: string | null;
  detection_method_id: number | null;
  failure_mode_id: number | null;
  failure_mechanism_id: number | null;
  failure_cause_id: number | null;
  status: string;
  occurred_at: string;
  detected_at: string;
  repair_started_at: string | null;
  resolved_at: string | null;
  title: string;
  description: string | null;
  root_cause_analysis: string | null;
  resolution_notes: string | null;
  man_hours: number | null;
  downtime_hours: number | null;
  operational_impact: string | null;
  reported_by_id: number;
  assigned_to_id: number | null;
  cabinet_name: string | null;
  reported_by_username: string | null;
  failure_mode_name: string | null;
  failure_mechanism_name: string | null;
  failure_cause_name: string | null;
  detection_method_name: string | null;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
};

export type MntIncidentComponent = {
  id: number;
  incident_id: number;
  cabinet_item_id: number;
  equipment_type_id: number | null;
  failure_mode_id: number | null;
  damage_description: string | null;
  action_taken: string | null;
  equipment_type_name: string | null;
  failure_mode_name: string | null;
  created_at: string;
};

export type MntWorkOrder = {
  id: number;
  order_number: string | null;
  order_type: string;
  activity_type_id: number | null;
  priority: string;
  status: string;
  cabinet_id: number | null;
  incident_id: number | null;
  plan_id: number | null;
  planned_start_date: string | null;
  planned_end_date: string | null;
  actual_start_at: string | null;
  actual_end_at: string | null;
  estimated_man_hours: number | null;
  actual_man_hours: number | null;
  assigned_to_id: number | null;
  performed_by_id: number;
  title: string;
  description: string | null;
  completion_notes: string | null;
  cabinet_name: string | null;
  activity_type_name: string | null;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
};

export type MntPlan = {
  id: number;
  name: string;
  code: string | null;
  equipment_category_id: number | null;
  equipment_type_id: number | null;
  cabinet_id: number | null;
  interval_days: number;
  activity_type_id: number | null;
  estimated_man_hours: number | null;
  description: string | null;
  last_generated_date: string | null;
  next_due_date: string | null;
  cabinet_name: string | null;
  activity_type_name: string | null;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
};

export type MntOperatingTime = {
  id: number;
  cabinet_id: number;
  recorded_date: string;
  operating_hours: number;
  standby_hours: number;
  downtime_hours: number;
  recorded_by_id: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
  cabinet_name: string | null;
};

export type ReliabilitySummary = {
  cabinet_id: number | null;
  cabinet_name: string | null;
  total_incidents: number;
  total_operating_hours: number;
  total_downtime_hours: number;
  mtbf_hours: number | null;
  mttr_hours: number | null;
  availability_pct: number | null;
};

export type FailureTrendPoint = {
  period: string;
  incident_count: number;
};

export type TopFailure = {
  equipment_type_id: number;
  equipment_type_name: string | null;
  incident_count: number;
};

// ---------------------------------------------------------------------------
// Dictionary API
// ---------------------------------------------------------------------------

const dictEndpoints = {
  failureModes: "/maintenance/failure-modes",
  failureMechanisms: "/maintenance/failure-mechanisms",
  failureCauses: "/maintenance/failure-causes",
  detectionMethods: "/maintenance/detection-methods",
  activityTypes: "/maintenance/activity-types",
} as const;

export function listMntDict(kind: keyof typeof dictEndpoints, params: Record<string, any> = {}) {
  return apiFetch<Pagination<MntDict>>(`${dictEndpoints[kind]}${buildQuery(params)}`);
}

export function createMntDict(kind: keyof typeof dictEndpoints, payload: any) {
  return apiFetch<MntDict>(dictEndpoints[kind], { method: "POST", body: JSON.stringify(payload) });
}

export function updateMntDict(kind: keyof typeof dictEndpoints, id: number, payload: any) {
  return apiFetch<MntDict>(`${dictEndpoints[kind]}/${id}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export function deleteMntDict(kind: keyof typeof dictEndpoints, id: number) {
  return apiFetch<void>(`${dictEndpoints[kind]}/${id}`, { method: "DELETE" });
}

export function restoreMntDict(kind: keyof typeof dictEndpoints, id: number) {
  return apiFetch<void>(`${dictEndpoints[kind]}/${id}/restore`, { method: "POST" });
}

// ---------------------------------------------------------------------------
// Incidents API
// ---------------------------------------------------------------------------

export function listIncidents(params: Record<string, any> = {}) {
  return apiFetch<Pagination<MntIncident>>(`/maintenance/incidents/${buildQuery(params)}`);
}

export function getIncident(id: number) {
  return apiFetch<MntIncident>(`/maintenance/incidents/${id}`);
}

export function createIncident(payload: any) {
  return apiFetch<MntIncident>("/maintenance/incidents/", { method: "POST", body: JSON.stringify(payload) });
}

export function updateIncident(id: number, payload: any) {
  return apiFetch<MntIncident>(`/maintenance/incidents/${id}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export function deleteIncident(id: number) {
  return apiFetch<void>(`/maintenance/incidents/${id}`, { method: "DELETE" });
}

export function restoreIncident(id: number) {
  return apiFetch<void>(`/maintenance/incidents/${id}/restore`, { method: "POST" });
}

export function listIncidentComponents(incidentId: number) {
  return apiFetch<MntIncidentComponent[]>(`/maintenance/incidents/${incidentId}/components`);
}

export function addIncidentComponent(incidentId: number, payload: any) {
  return apiFetch<MntIncidentComponent>(`/maintenance/incidents/${incidentId}/components`, {
    method: "POST", body: JSON.stringify(payload),
  });
}

export function deleteIncidentComponent(incidentId: number, componentId: number) {
  return apiFetch<void>(`/maintenance/incidents/${incidentId}/components/${componentId}`, { method: "DELETE" });
}

// ---------------------------------------------------------------------------
// Work Orders API
// ---------------------------------------------------------------------------

export function listWorkOrders(params: Record<string, any> = {}) {
  return apiFetch<Pagination<MntWorkOrder>>(`/maintenance/work-orders/${buildQuery(params)}`);
}

export function getWorkOrder(id: number) {
  return apiFetch<MntWorkOrder>(`/maintenance/work-orders/${id}`);
}

export function createWorkOrder(payload: any) {
  return apiFetch<MntWorkOrder>("/maintenance/work-orders/", { method: "POST", body: JSON.stringify(payload) });
}

export function updateWorkOrder(id: number, payload: any) {
  return apiFetch<MntWorkOrder>(`/maintenance/work-orders/${id}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export function deleteWorkOrder(id: number) {
  return apiFetch<void>(`/maintenance/work-orders/${id}`, { method: "DELETE" });
}

export function restoreWorkOrder(id: number) {
  return apiFetch<void>(`/maintenance/work-orders/${id}/restore`, { method: "POST" });
}

// ---------------------------------------------------------------------------
// Plans API
// ---------------------------------------------------------------------------

export function listPlans(params: Record<string, any> = {}) {
  return apiFetch<Pagination<MntPlan>>(`/maintenance/plans/${buildQuery(params)}`);
}

export function createPlan(payload: any) {
  return apiFetch<MntPlan>("/maintenance/plans/", { method: "POST", body: JSON.stringify(payload) });
}

export function updatePlan(id: number, payload: any) {
  return apiFetch<MntPlan>(`/maintenance/plans/${id}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export function deletePlan(id: number) {
  return apiFetch<void>(`/maintenance/plans/${id}`, { method: "DELETE" });
}

export function restorePlan(id: number) {
  return apiFetch<void>(`/maintenance/plans/${id}/restore`, { method: "POST" });
}

// ---------------------------------------------------------------------------
// Operating Time API
// ---------------------------------------------------------------------------

export function listOperatingTime(params: Record<string, any> = {}) {
  return apiFetch<Pagination<MntOperatingTime>>(`/maintenance/operating-time/${buildQuery(params)}`);
}

export function createOperatingTime(payload: any) {
  return apiFetch<MntOperatingTime>("/maintenance/operating-time/", { method: "POST", body: JSON.stringify(payload) });
}

export function updateOperatingTime(id: number, payload: any) {
  return apiFetch<MntOperatingTime>(`/maintenance/operating-time/${id}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export function deleteOperatingTime(id: number) {
  return apiFetch<void>(`/maintenance/operating-time/${id}`, { method: "DELETE" });
}

// ---------------------------------------------------------------------------
// Reliability API
// ---------------------------------------------------------------------------

export function getReliabilitySummary(params: Record<string, any> = {}) {
  return apiFetch<ReliabilitySummary[]>(`/maintenance/reliability/summary${buildQuery(params)}`);
}

export function getFailureTrend(params: Record<string, any> = {}) {
  return apiFetch<FailureTrendPoint[]>(`/maintenance/reliability/failure-trend${buildQuery(params)}`);
}

export function getTopFailures(params: Record<string, any> = {}) {
  return apiFetch<TopFailure[]>(`/maintenance/reliability/top-failures${buildQuery(params)}`);
}
